File: Data Structures & Algorithms/insert-into-a-binary-search-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def insertIntoBST(self, root: Optional[TreeNode], val: int) -> Optional[TreeNode]:
        if not root:
            return TreeNode(val)
        
        curr_node = root
        while True:
            if val > curr_node.val:
                if not curr_node.right:
                    curr_node.right = TreeNode(val)
                    return root
                curr_node = curr_node.right
            else:
                if not curr_node.left:
                    curr_node.left = TreeNode(val)
                    return root
                curr_node = curr_node.left
            

File: Data Structures & Algorithms/binary-tree-postorder-traversal/submission-2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<int> postorderTraversal(TreeNode* root) {
        vector<int> out;
        if (!root) return out;

        stack<pair<TreeNode*, bool>> s;
        s.push({root, false});
        while (!s.empty()) {
            TreeNode* curr = s.top().first;
            bool visited = s.top().second;
            s.pop();

            if (visited) {
                out.push_back(curr->val);
            } else {
                s.push({curr, true});
                if (curr->right) s.push({curr->right, false});
                if (curr->left) s.push({curr->left, false});
            }
        }
        return out;
    }
};